### ISSUE TITLE
Feature Request: async Shader compilation CPU affinity

### ISSUE THREAD

**Issue by phly95** [yuzu-emu/yuzu#6212]
Shader compilation is a demanding task. Asynchronous shader compilation alleviates some of the stutter, but what would be great is if shader compilation could be only done on cores separate from the game for high core count CPUs. This could eliminate any kind of shader compilation slowdown with the potential side effect of some effects taking longer to load in.

**Comment by goldenx86** [comment]
This is both CPU and OS dependent, it's not the responsibility of the program to handle it.

**Comment by phly95** [comment]
Maybe implement a shader compilation queue, where the user can choose how many shaders to compile at once? Basically users with more cores can have a higher number before it causes slowdown?

**Comment by ReinUsesLisp** [comment]
The stutter you see during shader compilation when using asynchronous shaders is because the initial decoding task is done on the main GPU thread.

Changing the affinity won't fix this. but I agree the shader worker threads should be marked as low priority to avoid wasting cycles in them when there's more meaningful stuff to do. On the new shader recompiler how shaders are compiled has changed to allow building them in parallel even when asynchronous shaders are off, so I'll mark the worker threads as low priority there.

As for the option to use more or less threads, that's already decided from the number of threads the device has.

**Comment by Morph1984** [comment]
> Maybe implement a shader compilation queue, where the user can choose how many shaders to compile at once? Basically users with more cores can have a higher number before it causes slowdown?
> 
> I do remember Cemuhook having an affinity option for cemu as a whole, though doing something like that for shaders may require a dedicated shader compilation process (as in Windows process).

Cemuhook's affinity option affects the whole program, and is exactly the same as setting the CPU Affinity in the Task Manager. However, you cannot set affinity for the individual threads in the program without a custom scheduler

**Comment by phly95** [comment]
Sorry if this is off-topic, but would resizable bar impact performance in any meaningful way?

**Comment by v1993** [comment]
> However, you cannot set affinity for the individual threads in the program without a custom scheduler

Are you sure? Quick search tells me that

https://docs.microsoft.com/en-us/windows/win32/procthread/multiple-processors:

> To set the thread affinity for a single thread, use the SetThreadAffinityMask function

https://man7.org/linux/man-pages/man3/pthread_setaffinity_np.3.html:

> pthread_setaffinity_np, pthread_getaffinity_np - set/get CPU
>        affinity of a thread

So it is supported both on Windows and on POSIX-compliant systems, even if interface is different.

**Comment by Morph1984** [comment]
Yea which is why I said a custom scheduler is needed, to make use of these platform specific mechanisms + we have to replace all instances of `std::thread` with a custom one since you *cant* do it with STL threads

**Comment by v1993** [comment]
I have not actually checked, but I'm fairly confident it is possible to use [`std::thread::native_handle()`](https://en.cppreference.com/w/cpp/thread/thread/native_handle) to access handle usable with said functions, thus combining them with STL threads.

**Comment by ReinUsesLisp** [comment]
We don't have to care about affinity. Setting the priority of the thread is enough, we have `Common::SetCurrentThreadPriority` for that task.

**Comment by german77** [comment]
request
